### PR TITLE
Pull out conv2d kernels from op

### PR DESCRIPTION
**Pull request by lockshaw**
By pulling out the kernel functions into their own files, we should decrease build time by reducing the number of times the kernels (which rarely change) need to be recompiled

### CODE REVIEW

File: src/ops/conv_2d.cc [hunk 1]
```diff
@@ -1074,6 +1081,96 @@ tl::optional<RecordFormatter> Conv2D::as_dot() const {
   return rr;
 }
 
+// TODO: refactor it
```
Comment (virena) [inline src/ops/conv_2d.cc:1084]: Is this still an issue? Or was the refactor to move it from the `.cu` file into this one?

**Review by virena — COMMENTED** [review]
Overall, it looks good to me. I have one question that's related to the codebase. The `.cpp` file seem to mimic its corresponding `.cu` file. Is it auto-generated? If not, should that also be modified for this PR? 

Also, what is the purpose of the kernel wrappers? If there's a `forward` and a `backward` and then also have `forward_kernel` and `backward_kernel`, I'm wondering if the wrapper is really necessary because it mostly seems to be doing some profiling.

File: src/ops/kernels/conv_2d_kernels.cu [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+    cudaEventDestroy(t_end);
+    printf("%s [Conv2D] backward time = %.2fms\n", m->op_name, elapsed);
+    // print_tensor<4, float>(acc_output_grad.ptr, acc_output_grad.rect,
+    // "[Conv2D:backward:output_grad]"); print_tensor<4,
+    // float>(acc_kernel_grad.ptr, acc_kernel_grad.rect,
+    // "[Conv2D:backward:kernel_grad]"); print_tensor<1,
+    // float>(acc_bias_grad.ptr, acc_bias_grad.rect,
+    // "[Conv2D:backward:bias_grad]"); print_tensor<4,
+    // float>(acc_input_grad.ptr, acc_input_grad.rect,
+    // "[Conv2D:backward:input_grad]");
```
Comment (virena) [inline src/ops/kernels/conv_2d_kernels.cu:232]: Remove?

Comment (jiazhihao) [reply src/ops/kernels/conv_2d_kernels.cu:232]: > The .cpp file seem to mimic its corresponding .cu file. Is it auto-generated? If not, should that also be modified for this PR?

The `.cpp` files are for running FlexFlow on AMD GPUs. They should be updated as well. 

Is this PR ready to be merged? I am about to push some changes to the fused.cc and fused.cu, which may conflict with some of your changes.

Comment (lockshaw) [reply src/ops/kernels/conv_2d_kernels.cu:232]: Not yet, go ahead and push those changes and I'll deal with any conflicts

**Comment by lockshaw** [comment]
Blocked on #345 as currently I can't check that the AMD kernel changes are correct

File: src/ops/conv_2d.cc (hunk 1 above)
Comment (lockshaw) [reply src/ops/conv_2d.cc:1086]: Honestly I don't remember so I'll get rid of it

**Review by gabrieleoliaro — APPROVED** [review]

**Review by virena — COMMENTED** [review]
What is the purpose of the change from `typename` to `class` in `model.h`? Otherwise LGTM

**Comment by lockshaw** [comment]
> What is the purpose of the change from `typename` to `class` in `model.h`? Otherwise LGTM

This is a C++17 feature apparently :shrug:, whereas `class` is `C++11` compliant. C++11 compliance checking is being added #271

**Pull request merged**